File: query/src/main/java/com/quarkworks/android/realmtypesafequery/fields/RealmByteField.java
package com.quarkworks.android.realmtypesafequery.fields;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.quarkworks.android.realmtypesafequery.BaseRealmField;
import com.quarkworks.android.realmtypesafequery.interfaces.SortableRealmField;

import io.realm.RealmModel;
import io.realm.RealmQuery;

public class RealmByteField<Model extends RealmModel> extends BaseRealmField<Model, Byte>
        implements SortableRealmField<Model, Byte> {

    public RealmByteField(@NonNull Class<Model> modelClass, @NonNull String keyPath) {
        super(modelClass, keyPath);
    }

    @Override
    public void equalTo(@NonNull RealmQuery<Model> query, @Nullable Byte value) {
        if (value == null) {
            this.isNull(query);
            return;
        }

        query.equalTo(this.getKeyPath(), value);
    }

    @Override
    public void notEqualTo(@NonNull RealmQuery<Model> query, @Nullable Byte value) {
        if (value == null) {
            this.isNotNull(query);
            return;
        }

        query.notEqualTo(this.getKeyPath(), value);
    }

    @Override
    public void never(@NonNull RealmQuery<Model> query) {
        query.equalTo(this.getKeyPath(), (byte) 0).equalTo(this.getKeyPath(), (byte) 1);
    }
}

File: query/src/main/java/com/quarkworks/android/realmtypesafequery/fields/RealmByteField.kt
package com.quarkworks.android.realmtypesafequery.fields

import io.realm.RealmModel
import io.realm.RealmQuery

open class RealmByteField<Model : RealmModel>(override val modelClass: Class<Model>, override val name: String) :
        RealmField<Model>,
        RealmEquatableField<Model, Byte>,
        RealmSortableField<Model, Byte>,
        RealmInableField<Model, Byte> {

    override fun equalTo(query: RealmQuery<Model>, value: Byte?) {
        if (value == null) {
            isNull(query)
            return
        }

        query.equalTo(name, value)
    }

    override fun notEqualTo(query: RealmQuery<Model>, value: Byte?) {
        if (value == null) {
            isNotNull(query)
            return
        }

        query.notEqualTo(name, value)
    }

    override fun never(query: RealmQuery<Model>) {
        query.beginGroup().equalTo(name, 0.toByte()).equalTo(name, 1.toByte()).endGroup()
    }

    override fun `in`(query: RealmQuery<Model>, values: Array<Byte>) {
        query.`in`(name, values)
    }
}
